src/partition/src/multi_dim.rs

```diff
@@ -134,6 +151,128 @@ impl MultiDimPartitionRule {
+        let mut current_row = vec![Value::Null; self.partition_columns.len()];
+        for row_idx in 0..num_rows {
+            self.row_at(&cols, row_idx, &mut current_row)?;
+            let current_region = self.find_region(&current_row)?;
+            let region_mask = result
+                .get_mut(&current_region)
+                .unwrap_or_else(|| panic!("Region {} must be initialized", current_region));
+            region_mask.set_bit(row_idx, true);
+        }
+
+        Ok(result
+            .into_iter()
+            .map(|(region, mut mask)| (region, BooleanArray::new(mask.finish(), None)))
+            .collect())
+    }
+
+    pub fn split_record_batch(
+        &self,
+        record_batch: &RecordBatch,
+    ) -> Result<HashMap<RegionNumber, BooleanArray>> {
+        let num_rows = record_batch.num_rows();
+        let physical_exprs = {
+            let cache_read_guard = self.physical_expr_cache.read().unwrap();
+            if let Some((cached_exprs, schema)) = cache_read_guard.as_ref()
+                && schema == record_batch.schema_ref()
+            {
+                cached_exprs.clone()
+            } else {
+                drop(cache_read_guard); // Release the read lock before acquiring write lock
+
+                let schema = record_batch.schema();
+                let new_cache = self
+                    .exprs
+                    .iter()
+                    .map(|e| e.try_as_physical_expr(&schema))
+                    .collect::<Result<Vec<_>>>()?;
+
+                let mut cache_write_guard = self.physical_expr_cache.write().unwrap();
+                cache_write_guard.replace((new_cache.clone(), schema));
+                new_cache
+            }
+        };
+
+        let mut result: HashMap<u32, BooleanArray> = physical_exprs
+            .iter()
+            .zip(self.regions.iter())
+            .map(|(expr, region_num)| {
+                let ColumnarValue::Array(column) = expr
+                    .evaluate(record_batch)
+                    .context(error::EvaluateRecordBatchSnafu)?
+                else {
+                    unreachable!("Expected an array")
+                };
+                Ok((
+                    *region_num,
+                    column
+                        .as_any()
+                        .downcast_ref::<BooleanArray>()
+                        .with_context(|| error::UnexpectedColumnTypeSnafu {
+                            data_type: column.data_type().clone(),
+                        })?
+                        .clone(),
+                ))
+            })
+            .collect::<error::Result<_>>()?;
+
+        let mut selected = BooleanArray::new(BooleanBuffer::new_unset(num_rows), None);
+        for region_selection in result.values() {
+            selected = arrow::compute::kernels::boolean::or(&selected, region_selection)
+                .context(error::ComputeArrowKernelSnafu)?;
+        }
```

Comment:
```suggestion
        }
        let all_rows_selected = selected.false_count() == 0;
        if all_rows_selected {
            return Ok(result);
        }
```

but how can a row not selected? I think it's expected that partition rule works on all rows.

Reply:
Note that in the row-wise `find_region` method, if a row is not selected by any parittion rule, it will be assigned to the default region:

https://github.com/GreptimeTeam/greptimedb/blob/62cd1f1f870ee6811ef760f34caf6ce0e7e66359/src/partition/src/multi_dim.rs#L91-L108

This behavior lacks a documentation [here](https://docs.greptime.com/user-guide/administration/manage-data/table-sharding/#create-a-distributed-table), in this case, we defined 4 partition rules:

```
PARTITION ON COLUMNS (device_id, area) (
  device_id < 100 AND area = 'North',
  device_id < 100 AND area = 'South',
  device_id >= 100 AND area = 'East',
  device_id >= 100 AND area = 'West'
);
```

And it will be natural for readers to ask: what if I insert a row with, for example, `device_id = 99` and `area = 'East'`?

Reply:
then the "amendment" here can be removed since the default region is already choosed inside